File: src/code-samples/langchain/deep-agent-from-scratch.py
"""Build a data analysis agent from scratch using create_agent and Deep Agents middleware."""

# :snippet-start: deep-agent-from-scratch-minimal-py
from langchain.agents import create_agent

agent = create_agent("anthropic:claude-sonnet-4-6", tools=[])
# :snippet-end:

# :snippet-start: deep-agent-from-scratch-sandbox-py
from langchain.agents import create_agent
from deepagents.backends.langsmith import LangSmithSandbox
from deepagents.middleware import FilesystemMiddleware
from langsmith.sandbox import SandboxClient

client = SandboxClient()
# :remove-start:
import atexit
import time

SANDBOX_NAME = "langchain-docs"


def _named_sandboxes() -> list[object]:
    return [sb for sb in client.list_sandboxes() if getattr(sb, "name", None) == SANDBOX_NAME]


def _delete_named_sandboxes() -> None:
    for existing in _named_sandboxes():
        for identifier in (
            getattr(existing, "sandbox_id", None),
            getattr(existing, "id", None),
            getattr(existing, "name", None),
            SANDBOX_NAME,
        ):
            if not identifier:
                continue
            try:
                client.delete_sandbox(identifier)
                break
            except Exception:
                continue


for _ in range(3):
    _delete_named_sandboxes()
    if not _named_sandboxes():
        break
    time.sleep(1)
# :remove-end:
sandbox = None
# :remove-start:

atexit.register(_delete_named_sandboxes)
# :remove-end:
sandbox = client.create_sandbox(name="langchain-docs", snapshot_name="docs-test-ci")
backend = LangSmithSandbox(sandbox=sandbox)

agent = create_agent(
    "anthropic:claude-sonnet-4-6",
    tools=[],
    middleware=[FilesystemMiddleware(backend=backend)],
)

# :snippet-end:

# :snippet-start: deep-agent-from-scratch-upload-py
import csv
import io

rows = [
    ["Date", "Product", "Units", "Revenue"],
    ["2025-08-01", "Widget A", 10, 250],
    ["2025-08-02", "Widget B", 5, 125],
    ["2025-08-03", "Widget A", 7, 175],
    ["2025-08-04", "Widget C", 3, 90],
]
buf = io.StringIO()
csv.writer(buf).writerows(rows)
backend.upload_files([("/sales.csv", buf.getvalue().encode())])

upload_stream = agent.stream_events(
    {
        "messages": [
            {
                "role": "user",
                "content": (
                    "Read /sales.csv and summarize total revenue by product in one "
                    "sentence. Do not run shell commands."
                ),
            }
        ]
    },
    version="v3",
    config={"recursion_limit": 8},
)
for item in upload_stream.messages:
    print(item.text)
upload_stream.output
# :snippet-end:

# :snippet-start: deep-agent-from-scratch-summarization-py
from deepagents.middleware import FilesystemMiddleware, SummarizationMiddleware

model = "anthropic:claude-sonnet-4-6"

agent = create_agent(
    model=model,
    tools=[],
    middleware=[
        FilesystemMiddleware(backend=backend),
        SummarizationMiddleware(model=model, backend=backend),
    ],
)
# :snippet-end:

# :snippet-start: deep-agent-from-scratch-skills-upload-py
from pathlib import Path

skills_dir = (Path(__file__).resolve().parent / "skills").resolve()
# :remove-start:
skills_dir = Path("src/code-samples/langchain/skills").resolve()
# :remove-end:
skill_files: list[tuple[str, bytes]] = []
for path in sorted(skills_dir.rglob("*")):
    if not path.is_file():
        continue
    rel = path.resolve().relative_to(skills_dir)
    skill_files.append((f"/skills/{rel.as_posix()}", path.read_bytes()))
backend.upload_files(skill_files)
# :snippet-end:

# :snippet-start: deep-agent-from-scratch-skills-py
from deepagents.middleware import FilesystemMiddleware, SkillsMiddleware, SummarizationMiddleware

agent = create_agent(
    model=model,
    tools=[],
    middleware=[
        FilesystemMiddleware(backend=backend),
        SummarizationMiddleware(model=model, backend=backend),
        SkillsMiddleware(backend=backend, sources=["/skills/"]),
    ],
)
# :snippet-end:

# :snippet-start: deep-agent-from-scratch-subagent-py
from deepagents import SubAgent
from deepagents.middleware import (
    FilesystemMiddleware,
    SkillsMiddleware,
    SubAgentMiddleware,
    SummarizationMiddleware,
)
from langchain.agents.middleware import TodoListMiddleware

visualizer: SubAgent = {
    "name": "visualizer",
    "description": "Generates charts and visualizations from data files in the sandbox.",
    "system_prompt": "You are a data visualization specialist. Write Python scripts using matplotlib and seaborn. Save all figures as PNG files.",
    "tools": [],
    "model": "anthropic:claude-sonnet-4-6",
}

agent = create_agent(
    model=model,
    tools=[],
    middleware=[
        FilesystemMiddleware(backend=backend),
        SummarizationMiddleware(model=model, backend=backend),
        SkillsMiddleware(backend=backend, sources=["/skills/"]),
        TodoListMiddleware(),
        SubAgentMiddleware(backend=backend, subagents=[visualizer]),
    ],
)
# :snippet-end:

# :remove-start:
try:
    assert backend.read("/sales.csv").error is None
    assert backend.read("/skills/pandas-patterns/SKILL.md").error is None
    assert agent is not None
    print("✓ deep-agent-from-scratch sample completed")
finally:
    _delete_named_sandboxes()
# :remove-end:

File: src/code-samples/langchain/deep-agent-from-scratch.ts
/** Build a data analysis agent from scratch using createAgent and Deep Agents middleware. */

// :snippet-start: deep-agent-from-scratch-minimal-js
import { createAgent } from "langchain";

let agent = createAgent({
  model: "openai:gpt-4.1",
  tools: [],
});
// :snippet-end:

// :snippet-start: deep-agent-from-scratch-sandbox-js
import { createFilesystemMiddleware, LangSmithSandbox } from "deepagents";
import { SandboxClient } from "langsmith/sandbox";

const client = new SandboxClient();
// :remove-start:
const SANDBOX_NAME = "langchain-docs";

function sandboxIdentifiers(sb: unknown): Array<string> {
  const record = sb as Record<string, unknown>;
  return [record.sandboxId, record.id, record.name, SANDBOX_NAME].filter(
    (value): value is string => typeof value === "string" && value.length > 0,
  );
}

async function namedSandboxes() {
  const sandboxes = await client.listSandboxes();
  return Array.from(sandboxes).filter(
    (sb) => (sb as { name?: string }).name === SANDBOX_NAME,
  );
}

async function deleteNamedSandboxes() {
  for (const existing of await namedSandboxes()) {
    for (const identifier of sandboxIdentifiers(existing)) {
      try {
        await client.deleteSandbox(identifier);
        break;
      } catch {
        // Try the next identifier form.
      }
    }
  }
}

for (let attempt = 0; attempt < 3; attempt += 1) {
  await deleteNamedSandboxes();
  if ((await namedSandboxes()).length === 0) {
    break;
  }
  await new Promise((resolve) => setTimeout(resolve, 1000));
}
// :remove-end:
const sandbox = await client.createSandbox({
  name: "langchain-docs",
  snapshotName: "docs-test-ci",
});
// :remove-start:
process.once("exit", () => {
  void deleteNamedSandboxes();
});
// :remove-end:
const backend = new LangSmithSandbox({ sandbox });

agent = createAgent({
  model: "openai:gpt-4.1",
  tools: [],
  middleware: [createFilesystemMiddleware({ backend })],
});
// :snippet-end:

// :snippet-start: deep-agent-from-scratch-upload-js
const rows = [
  ["Date", "Product", "Units", "Revenue"],
  ["2025-08-01", "Widget A", "10", "250"],
  ["2025-08-02", "Widget B", "5", "125"],
  ["2025-08-03", "Widget A", "7", "175"],
  ["2025-08-04", "Widget C", "3", "90"],
];

const csv = rows.map((row) => row.join(",")).join("\n");
const encoder = new TextEncoder();
await backend.uploadFiles([["/sales.csv", encoder.encode(csv)]]);

const uploadStream = await agent.streamEvents(
  {
    messages: [
      {
        role: "user",
        content:
          "Read /sales.csv and summarize total revenue by product in one sentence. Do not run shell commands.",
      },
    ],
  },
  { version: "v3", recursionLimit: 8 },
);

await Promise.all([
  (async () => {
    for await (const message of uploadStream.messages) {
      console.log(await message.text);
    }
  })(),
  uploadStream.output,
]);
// :snippet-end:

// :snippet-start: deep-agent-from-scratch-summarization-js
import { createSummarizationMiddleware } from "deepagents";

agent = createAgent({
  model: "openai:gpt-4.1",
  tools: [],
  middleware: [
    createFilesystemMiddleware({ backend }),
    createSummarizationMiddleware({
      model: "openai:gpt-4.1",
      backend,
    }),
  ],
});
// :snippet-end:

// :snippet-start: deep-agent-from-scratch-skills-upload-js
import { readFileSync, readdirSync, statSync } from "node:fs";
import { join, relative, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const skillsDir = resolve(
  fileURLToPath(new URL(".", import.meta.url)),
  "skills",
);
const skillFiles: Array<[string, Uint8Array]> = [];

function collectSkillFiles(dir: string): void {
  for (const entry of readdirSync(dir)) {
    const fullPath = join(dir, entry);
    if (statSync(fullPath).isDirectory()) {
      collectSkillFiles(fullPath);
    } else {
      const rel = relative(skillsDir, fullPath).replace(/\\/g, "/");
      skillFiles.push([`/skills/${rel}`, readFileSync(fullPath)]);
    }
  }
}

collectSkillFiles(skillsDir);
await backend.uploadFiles(skillFiles);
// :snippet-end:

// :snippet-start: deep-agent-from-scratch-skills-js
import { createSkillsMiddleware } from "deepagents";

let model = "openai:gpt-4.1";

agent = createAgent({
  model,
  tools: [],
  middleware: [
    createFilesystemMiddleware({ backend }),
    createSummarizationMiddleware({ model, backend }),
    createSkillsMiddleware({ backend, sources: ["/skills/"] }),
  ],
});
// :snippet-end:

// :snippet-start: deep-agent-from-scratch-subagent-js
import { todoListMiddleware } from "langchain";
import { createSubAgentMiddleware, type SubAgent } from "deepagents";

const visualizer: SubAgent = {
  name: "visualizer",
  description:
    "Generates charts and visualizations from data files in the sandbox.",
  systemPrompt:
    "You are a data visualization specialist. Write Python scripts using matplotlib and seaborn. Save all figures as PNG files.",
  tools: [],
  model: "openai:gpt-4.1",
};

agent = createAgent({
  model,
  tools: [],
  middleware: [
    createFilesystemMiddleware({ backend }),
    createSummarizationMiddleware({ model, backend }),
    createSkillsMiddleware({ backend, sources: ["/skills/"] }),
    todoListMiddleware(),
    createSubAgentMiddleware({
      defaultModel: model,
      defaultTools: [],
      subagents: [visualizer],
    }),
  ],
});
// :snippet-end:

// :remove-start:
try {
  const salesRead = await backend.read("/sales.csv");
  if (salesRead.error) {
    throw new Error(salesRead.error);
  }
  const skillsRead = await backend.read("/skills/pandas-patterns/SKILL.md");
  if (skillsRead.error) {
    throw new Error(skillsRead.error);
  }
  if (!agent) {
    throw new Error("expected agent");
  }
  console.log("✓ deep-agent-from-scratch sample completed");
} finally {
  await deleteNamedSandboxes();
}
// :remove-end:
